utxo.go

```diff
@@ -0,0 +1,193 @@
+	// no recipient is set to receive max amount.
+	nextInternalAddress := func() (string, error) {
+		ctx := tx.sourceWallet.shutdownContext()
+		addr, err := tx.sourceWallet.internal.NewChangeAddress(ctx, tx.sourceAccountNumber)
+		if err != nil {
+			return "", err
+		}
+		return addr.Address(), nil
+	}
+
+	return tx.newUnsignedTxUTXO(tx.inputs, tx.destinations, tx.changeDestinations, nextInternalAddress)
+}
+
+func (tx *TxAuthor) newUnsignedTxUTXO(inputs []*wire.TxIn, sendDestinations, changeDestinations []TransactionDestination,
+	nextInternalAddress nextAddressFunc) (*txauthor.AuthoredTx, error) {
+	outputs, totalSendAmount, maxAmountRecipientAddress, err := tx.ParseOutputsAndChangeDestination(sendDestinations)
+	if err != nil {
+		return nil, err
+	}
+
+	if maxAmountRecipientAddress != "" && len(changeDestinations) > 0 {
+		return nil, errors.E(errors.Invalid, "no change is generated when sending max amount,"+
+			" change destinations must not be provided")
+	}
+
+	if maxAmountRecipientAddress == "" && len(changeDestinations) == 0 {
+		// no change specified, generate new internal address to use as change (max amount recipient)
+		maxAmountRecipientAddress, err = nextInternalAddress()
+		if err != nil {
+			return nil, fmt.Errorf("error generating internal address to use as change: %s", err.Error())
+		}
+	}
+
+	var totalInputAmount int64
+	inputScriptSizes := make([]int, len(inputs))
+	inputScripts := make([][]byte, len(inputs))
+	for i, input := range inputs {
+		totalInputAmount += input.ValueIn
+		inputScriptSizes[i] = txsizes.RedeemP2PKHSigScriptSize
+		inputScripts[i] = input.SignatureScript
+	}
+
+	var changeScriptSize int
+	if maxAmountRecipientAddress != "" {
+		changeScriptSize, err = calculateChangeScriptSize(maxAmountRecipientAddress, tx.sourceWallet.chainParams)
+	} else {
+		changeScriptSize, err = calculateMultipleChangeScriptSize(changeDestinations, tx.sourceWallet.chainParams)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	maxSignedSize := txsizes.EstimateSerializeSize(inputScriptSizes, outputs, changeScriptSize)
+	maxRequiredFee := txrules.FeeForSerializeSize(txrules.DefaultRelayFeePerKb, maxSignedSize)
+	changeAmount := totalInputAmount - totalSendAmount - int64(maxRequiredFee)
+
+	if changeAmount < 0 {
+		return nil, errors.New(ErrInsufficientBalance)
+	}
+
+	if changeAmount != 0 && !txrules.IsDustAmount(dcrutil.Amount(changeAmount), changeScriptSize, txrules.DefaultRelayFeePerKb) {
+		if changeScriptSize > txscript.MaxScriptElementSize {
+			return nil, fmt.Errorf("script size exceed maximum bytes pushable to the stack")
+		}
+		var totalChangeAmount int64
+		if maxAmountRecipientAddress != "" {
+			outputs, err = tx.changeOutput(changeAmount, maxAmountRecipientAddress, outputs)
+			if err != nil {
+				return nil, fmt.Errorf("change address error: %v", err)
+			}
+			totalChangeAmount = changeAmount
+		} else if len(changeDestinations) > 0 {
+			changeDestinations[0].AtomAmount = changeAmount
```

Comment:
Why is the entire change amount assigned to the first change destination?, this means there'll be no change left for the remaining change destinations.